[Runtime] Don't use threading APIs when building single threaded.

al45tair opened the pull request:
Mostly this consists of `#if`-ing out code that we don't need when running single threaded.

rdar://84393438

al45tair committed:
[Runtime] Don't use threading related APIs when building single threaded.

Mostly this consists of #if-ing out code that we don't need when running
single threaded.

rdar://84393438

al45tair commented:
@swift-ci Please smoke test

mikeash commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -559,7 +561,9 @@ void TaskGroupImpl::offer(AsyncTask *completedTask, AsyncContext *context) {
   SWIFT_TASK_DEBUG_LOG("offer task %p to group %p", completedTask, this);
 
+#if !SWIFT_STDLIB_SINGLE_THREADED_RUNTIME
   mutex.lock(); // TODO: remove fragment lock, and use status for synchronization
```
It would be nice to centralize these conditionals with small `lockMutex` and `unlockMutex` methods that are empty in the single-threaded runtime. That doesn't have to be done now, just something to consider if you find yourself bored one day.

al45tair replied:
I've thought that every time I looked at this changeset, but needed a spare moment to actually make the change. Will do that now.

mikeash replied:
Actually, `include/swift/Runtime/Mutex.h` has wrappers that will do the right thing for any given platform, and it already has a single-threaded stub implementation.

mikeash replied:
Yeah, not sure what the deal is with that. I'd expect `Mutex.h` to be a better choice at least on Darwin, where it will use `os_unfair_lock`, but it could be that there are weird problems with it in this context.

grynspan replied:
Does the C++11 standard (or later) allow for a conforming implementation that does not have `std::mutex`? I'd expect a single-threaded system to have the type but that using it would be a no-op.

al45tair replied:
@grynspan Regardless of what the C++ standard has to say on the subject, there are definitely systems (that we care about) where we can't use `std::mutex`.

al45tair committed:
[Runtime] Tidy up locking/unlocking conditionals.

Rather than conditionalising all the calls to mutex.lock() and mutex.unlock(),
centralise the logic.

rdar://84393438

al45tair commented:
@swift-ci Please smoke test

kubamracek commented:
Looks great. Can you make sure PR testing for the "minimal" preset passes before merging? That can be done via:
```
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset
```
But unfortunately right now that CI job is failing because we are waiting for a new nightly toolchain to be produced.

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

al45tair commented:
@swift-ci please smoke test windows platform

kubamracek commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

kubamracek commented:
@swift-ci please test Windows platform

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

al45tair commented:
@swift-ci please test Windows platform

al45tair committed:
[Tests] Add swift_Concurrency to the LTO workaround.

Kuba reckons this will fix the test hanging.

rdar://84393438

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

al45tair commented:
@swift-ci Please smoke test
